### ISSUE TITLE
Add extension object to ExtensionContext

### ISSUE THREAD

**Issue by eamodio** [microsoft/vscode#116906]
It would be great to expose both `extensionId` and `extensionVersion` on our `ExtensionContext`. I've seen extensions (too many times) actually `require('./package.json)' to get at either the extension id or version. We were (still are until a PR lands) even doing this in our GHPR extension. In the best case, this wastes cycles, causes I/O contention, slows activations, and in the worse (e.g. bundling) actually bundles the `package.json` into your JS bundle.

By at least providing access to the `id` (and the mostly commonly accessed `package.json` property `version`), extensions can then use the `extensions.getExtension()` api to get at the other `package.json` properties that we already cache.

```ts
	export interface ExtensionContext {
		readonly extensionId: string;
		readonly extensionVersion: string;
	}
```

/cc @jrieken @sandy081

**Comment by jrieken** [comment]
I find it funny that an extension does not know its own id? With that in mind you can go `vscode.extensions.getExtension('me.my')` and the returned object has everything you need. 

I also don't see the logic in cherry-picking some properties and exposing/re-exporting them like that. Why version and not publisher or dependencies. If we really think knowing your own extension id is too much of a burden than we should simply put the whole `Extension` object on the context

**Comment by eamodio** [comment]
So an extension can not know it's own id when creating an insiders/nightly version from the same source, e.g. GHPR, GitLens, JS-Debug, etc. 

As for cherry picking only those 2 properties, the id gives you access to all the rest of needed, and I added version for convenience since that is commonly accessed and changes (sometimes through a build process) where as the other properties don't. 

Although I'd be fine with exposing the whole thing too. I almost added it, but figured there would be more objections to it 😂.

**Comment by jrieken** [comment]
> So an extension can not know it's own id when creating an insiders/nightly version from the same source, e.g. GHPR, GitLens, JS-Debug, etc.

I believe they still know their id. I don't know the code but I would assume there is some script that patches this onto its own package.json file before publishing. So, my question would be why that cannot done to its own sources.

**Comment by eamodio** [comment]
In the cases I've laid out, the running code *doesn't* know its id -- it has to look it up from somewhere or compare possible ids to the extension path (the former is what GHPR does by requiring the package.json, and the latter is what I do in GitLens to avoid loading the package.json).

And yes, technically a build script could update some runnable code or something in addition to the package.json places, but why deal with that when we VERY easily can expose it on the `ExtensionContext` where we already readily have that data.

**Comment by TylerLeonhardt** [comment]
Déjà vu ... https://github.com/microsoft/vscode/issues/66515

tl;dr: PowerShell extension had to require the package.json to get this info and it was needed because the PowerShell extension ships a "preview" extension and a "stable" extension in the marketplace. That sucks for perf (more so for GitLens & GHPR)

I'd love to see at least the `extensionId` in there for the `vscode.extensions.getExtension(extensionId)` scenario.

This would help make extension authorship of multiple extensions that use the same code (aka beta extensions) easier.

**Comment by jrieken** [comment]
> So why force/deal with that when we VERY easily can expose it on the ExtensionContext where we already readily have that data.

Please calm down, I am simply asking questions. This is clearly for a niche use-case for a handful of extension that might be obsolete as soon as the marketplace supports insider versions of extension. This use-case already already requires patching your sources and I don't think it is outrageous to wonder why that cannot be extended, esp. since the patch won't go away with this. 

When it comes to API the question isn't whether it is easy or very easy to implement. The question is what's the maintenance cost for an API (low in this case), what's the story for an API (exposing just two selected properties that happen to match your use-case is illogical), does the API duplicate an existing API (yes in this case), and so on. The goal of the API is to be consistent and not be the result of ad-hoc requests.

**Comment by TylerLeonhardt** [comment]
I think @lramos15 also requires the package.json to get his AppInsights id. This would also help that scenario because with the `extensionId` you can get your extension's `Extension` reference and then the packageJson already loaded in memory rather than loading it again just for that detail.

I think their are additional scenarios where information in the `package.json` might be valuable for the extension author and giving them access to the packageJson already in memory that we have is nicer than them having to `require` it again.

I understand that that assumes they don't know know their own extensionId and thus can't get it with `vscode.extensions.getExtension("my.ext")` and beyond the beta extensions use-case, I'm not sure there are others who don't know the extensionId

> extension that might be obsolete as soon as the marketplace supports insider versions of extension

I just want to say that this is one of the reasons my issue was closed 2 years ago. Here we are without beta releases still. I have some doubt that that will change in the next 2 to be perfectly honest. I hope I'm proven wrong.

tbh, I think I like adding a reference to the `Extension` object on `ExtensionContext`:

```ts
export interface ExtensionContext {
	readonly extension: Extension;
}
```

It plays nicely into "the `ExtensionContext` gives you all the context you need to run your extension - Including the details of the extension that you're trying to run."

**Comment by RMacfarlane** [comment]
> This use-case already already requires patching your sources and I don't think it is outrageous to wonder why that cannot be extended, esp. since the patch won't go away with this.

The GHPRI extension does in fact update a constant in a file with the id for this purpose, as well as changing the value in the `package.json` as part of the build. I forgot we did that since we haven't touched the build pipeline in years now so accidentally reintroduced reading the package.json recently

> I have some doubt that that will change in the next 2 to be perfectly honest. 

I feel actually somewhat optimistic about this since there now seems to be more communication from the marketplace team and this has made it onto their roadmap for the year

**Comment by lramos15** [comment]
From what I'm interpreting from @jrieken the best solution that isn't specialized and is easy to maintain comes down to just providing the extension object in the extension context. That way we're not exposing certain properties for some and not for others, this also allows me to get all the information about my extension without having to waste additional I/O cycles on reading the package.json that is already read in by VS Code. As @TylerLeonhardt points out many extensions use the package.json as a config bag because there are already extension related information in there and therefore each extension might want something else from it. To me it seems like the flow is often going to be:
1. Get the extension id from the context
2. Call the extension API to get the extension object
3. Do something with that

So we might as well remove steps 1 and 2 and just give them the entire extension object.

**Comment by jrieken** [comment]
> From what I'm interpreting from @jrieken the best solution that isn't specialized and is easy to maintain comes down to just providing the extension object in the extension context

Spot on. From my point of view knowing your extension id isn't an unmanageable burden but I can see how many of these calls might look cumbersome. Tho, at least it's globally available and doesn't require you to pass the context/description around. There are pros and cons

**Comment by jrieken** [comment]
Keeping this open because it isn't in `vscode.d.ts` yet
